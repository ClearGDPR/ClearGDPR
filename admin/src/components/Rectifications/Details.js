import React from 'react';
import PropTypes from 'prop-types';
import lodash from 'lodash';

import { rectificationDetailsType } from 'types';

import Loader from 'components/core/cards/dashboard/Loader';
import { PrimaryButton } from 'components/core/Common/Buttons/Buttons';

const Details = ({ rectification, onApprove, isLoading }) => {
  function renderObject(obj, level = 0) {
    if (!obj) return;

    return Object.keys(obj).map((k, index) => (
      <React.Fragment key={index}>
        {level ? <br /> : ''}
        {[...Array(level)].map((value, arrayIndex) => (
          <React.Fragment key={arrayIndex}>&nbsp;&nbsp;</React.Fragment>
        ))}
        {k}:&nbsp;
        {lodash.isObject(obj[k]) ? renderObject(obj[k], level + 1) : obj[k]}
        {!level ? <br /> : ''}
      </React.Fragment>
    ));
  }

  function renderDetails() {
    return (
      <React.Fragment>
        <p>
          <strong>Created at:</strong> {rectification.created_at}
        </p>
        <p>
          <strong>Current data:</strong>
        </p>
        <p>{renderObject(rectification.currentData)}</p>
        <p>
          <strong>Updates:</strong>
        </p>
        <p>{renderObject(rectification.updates)}</p>
        <p>Status: {rectification.status}</p>
        {rectification.status !== 'APPROVED' && (
          <PrimaryButton
            type="button"
            onClick={e => {
              e.preventDefault();
              onApprove();
            }}
            text="Approve"
          />
        )}
      </React.Fragment>
    );
  }
  return isLoading ? <Loader /> : renderDetails();
};

Details.propTypes = {
  rectification: rectificationDetailsType.isRequired,
  onApprove: PropTypes.func.isRequired,
  isLoading: PropTypes.bool
};

export default Details;
