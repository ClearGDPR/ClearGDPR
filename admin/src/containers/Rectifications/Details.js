import React from 'react';
import PropTypes from 'prop-types';
import format from 'helpers/date-formatter';

import config from 'config';
import internalFetch from 'helpers/internal-fetch';

import { RectificationsConsumer } from './RectificationsContext';
import { PanelConsumer } from 'containers/MainLayout/PanelContext';
import Details from 'components/Rectifications/Details';

export class DetailsContainer extends React.Component {
  static propTypes = {
    approveRectification: PropTypes.func,
    rectificationId: PropTypes.number,
    closePanel: PropTypes.func,
    isLoading: PropTypes.bool
  };

  state = {
    rectification: {}
  };

  componentDidMount() {
    this.setState({
      isLoading: true
    });

    internalFetch(
      `${config.API_URL}/api/management/subjects/rectification-requests/${
        this.props.rectificationId
      }`
    )
      .then(r =>
        this.setState({
          rectification: {
            created_at: format(r.createdAt),
            currentData: r.currentData,
            updates: r.updates,
            status: r.status
          },
          isLoading: false
        })
      )
      .catch(() => this.setState({ isLoading: false }));
  }

  get isLoading() {
    return this.state.isLoading || this.props.isLoading;
  }

  onApprove = () => {
    this.setState({
      isLoading: true
    });

    this.props
      .approveRectification(this.props.rectificationId)
      .then(() => this.props.closePanel())
      .catch(() => {
        this.setState({
          isLoading: false
        });
      });
  };

  render() {
    return (
      <Details
        rectification={this.state.rectification}
        onApprove={this.onApprove}
        isLoading={this.isLoading}
      />
    );
  }
}

export default props => (
  <PanelConsumer>
    {({ closePanel }) => (
      <RectificationsConsumer>
        {({ approveRectification, isLoading }) => (
          <DetailsContainer
            {...props}
            isLoading={isLoading}
            approveRectification={approveRectification}
            closePanel={closePanel}
          />
        )}
      </RectificationsConsumer>
    )}
  </PanelConsumer>
);
