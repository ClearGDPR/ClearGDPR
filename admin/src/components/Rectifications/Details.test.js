import React from 'react';
import { shallow, mount } from 'enzyme';

import Details from './Details';

const basicRectification = {
  id: 1,
  currentData: {
    firstName: 'Johnny',
    email: 'jb@coolcartoons.io',
    someKey: 'Some data goes here'
  },
  updates: {
    lastName: 'Bravo'
  },
  createdAt: '2018-07-10T16:14:20.742Z',
  status: 'PENDING'
};

const nestedRectification = {
  id: 2,
  currentData: {
    firstName: 'Johnny',
    email: 'jb@coolcartoons.io'
  },
  updates: {
    lastName: 'Bravo',
    someKey: 'Some data goes here',
    specifics: {
      hairStyle: 'fashionable',
      favoriteColor: 'black',
      moreSpecifics: {
        quote: "Man, I'm pretty."
      }
    }
  },
  createdAt: '2018-07-10T16:14:20.742Z',
  status: 'APPROVED'
};

const nestedRectificationWithArray = {
  id: 2,
  currentData: {
    firstName: 'Johnny',
    email: 'jb@coolcartoons.io',
    someArray: [1, 2, 'three']
  },
  updates: {
    lastName: 'Bravo',
    someKey: 'Some data goes here',
    specifics: {
      hairStyle: 'fashionable',
      favoriteColor: 'black',
      moreSpecifics: {
        quote: "Man, I'm pretty."
      },
      wardrobe: ['jeans', 't-shirt']
    }
  },
  createdAt: '2018-07-10T16:14:20.742Z',
  status: 'APPROVED'
};

const setup = propOverrides => {
  const props = Object.assign(
    {
      rectification: basicRectification,
      onApprove: jest.fn(),
      isLoading: false
    },
    propOverrides
  );
  const component = shallow(<Details {...props} />);
  const mounted = mount(<Details {...props} />);

  return { props, component, mounted };
};

beforeEach(() => {
  jest.resetAllMocks();
});

describe('(Component) Rectification details', () => {
  it('should render when not loading correctly', () => {
    const { component } = setup();
    expect(component).toMatchSnapshot();
  });

  it('should render when loading correctly', () => {
    const component = setup({ isLoading: true });
    expect(component).toMatchSnapshot();
  });

  it('should render nested properties correctly', () => {
    const { component, mounted } = setup({ rectification: nestedRectification });
    expect(component).toMatchSnapshot();

    const button = mounted.find('button').at(0);
    expect(button.exists()).toBe(false);
  });

  it('should render array values in properties correctly', () => {
    const { component } = setup({ rectification: nestedRectificationWithArray });
    expect(component).toMatchSnapshot();
  });

  it('should call onApproved on click of Approve button', () => {
    const { props, mounted } = setup();
    const button = mounted.find('button').at(0);
    button.simulate('click');

    expect(props.onApprove).toHaveBeenCalled();
  });
});
