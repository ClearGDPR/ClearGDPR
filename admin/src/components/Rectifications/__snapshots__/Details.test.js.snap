// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Rectification details should render array values in properties correctly 1`] = `
<React.Fragment>
  <p>
    <strong>
      Created at:
    </strong>
     
  </p>
  <p>
    <strong>
      Current data:
    </strong>
  </p>
  <p>
    <React.Fragment
      key="0"
    >
      firstName
      : 
      Johnny
      <br />
    </React.Fragment>
    <React.Fragment
      key="1"
    >
      email
      : 
      jb@coolcartoons.io
      <br />
    </React.Fragment>
    <React.Fragment
      key="2"
    >
      someArray
      : 
      <React.Fragment
        key="0"
      >
        <br />
        <React.Fragment
          key="0"
        >
            
        </React.Fragment>
        0
        : 
        1
      </React.Fragment>
      <React.Fragment
        key="1"
      >
        <br />
        <React.Fragment
          key="0"
        >
            
        </React.Fragment>
        1
        : 
        2
      </React.Fragment>
      <React.Fragment
        key="2"
      >
        <br />
        <React.Fragment
          key="0"
        >
            
        </React.Fragment>
        2
        : 
        three
      </React.Fragment>
      <br />
    </React.Fragment>
  </p>
  <p>
    <strong>
      Updates:
    </strong>
  </p>
  <p>
    <React.Fragment
      key="0"
    >
      lastName
      : 
      Bravo
      <br />
    </React.Fragment>
    <React.Fragment
      key="1"
    >
      someKey
      : 
      Some data goes here
      <br />
    </React.Fragment>
    <React.Fragment
      key="2"
    >
      specifics
      : 
      <React.Fragment
        key="0"
      >
        <br />
        <React.Fragment
          key="0"
        >
            
        </React.Fragment>
        hairStyle
        : 
        fashionable
      </React.Fragment>
      <React.Fragment
        key="1"
      >
        <br />
        <React.Fragment
          key="0"
        >
            
        </React.Fragment>
        favoriteColor
        : 
        black
      </React.Fragment>
      <React.Fragment
        key="2"
      >
        <br />
        <React.Fragment
          key="0"
        >
            
        </React.Fragment>
        moreSpecifics
        : 
        <React.Fragment
          key="0"
        >
          <br />
          <React.Fragment
            key="0"
          >
              
          </React.Fragment>
          <React.Fragment
            key="1"
          >
              
          </React.Fragment>
          quote
          : 
          Man, I'm pretty.
        </React.Fragment>
      </React.Fragment>
      <React.Fragment
        key="3"
      >
        <br />
        <React.Fragment
          key="0"
        >
            
        </React.Fragment>
        wardrobe
        : 
        <React.Fragment
          key="0"
        >
          <br />
          <React.Fragment
            key="0"
          >
              
          </React.Fragment>
          <React.Fragment
            key="1"
          >
              
          </React.Fragment>
          0
          : 
          jeans
        </React.Fragment>
        <React.Fragment
          key="1"
        >
          <br />
          <React.Fragment
            key="0"
          >
              
          </React.Fragment>
          <React.Fragment
            key="1"
          >
              
          </React.Fragment>
          1
          : 
          t-shirt
        </React.Fragment>
      </React.Fragment>
      <br />
    </React.Fragment>
  </p>
  <p>
    Status: 
    APPROVED
  </p>
  <button
    className="btn ui-action"
    onClick={[Function]}
    type="button"
  >
    Approve
  </button>
</React.Fragment>
`;

exports[`(Component) Rectification details should render nested properties correctly 1`] = `
<React.Fragment>
  <p>
    <strong>
      Created at:
    </strong>
     
  </p>
  <p>
    <strong>
      Current data:
    </strong>
  </p>
  <p>
    <React.Fragment
      key="0"
    >
      firstName
      : 
      Johnny
      <br />
    </React.Fragment>
    <React.Fragment
      key="1"
    >
      email
      : 
      jb@coolcartoons.io
      <br />
    </React.Fragment>
  </p>
  <p>
    <strong>
      Updates:
    </strong>
  </p>
  <p>
    <React.Fragment
      key="0"
    >
      lastName
      : 
      Bravo
      <br />
    </React.Fragment>
    <React.Fragment
      key="1"
    >
      someKey
      : 
      Some data goes here
      <br />
    </React.Fragment>
    <React.Fragment
      key="2"
    >
      specifics
      : 
      <React.Fragment
        key="0"
      >
        <br />
        <React.Fragment
          key="0"
        >
            
        </React.Fragment>
        hairStyle
        : 
        fashionable
      </React.Fragment>
      <React.Fragment
        key="1"
      >
        <br />
        <React.Fragment
          key="0"
        >
            
        </React.Fragment>
        favoriteColor
        : 
        black
      </React.Fragment>
      <React.Fragment
        key="2"
      >
        <br />
        <React.Fragment
          key="0"
        >
            
        </React.Fragment>
        moreSpecifics
        : 
        <React.Fragment
          key="0"
        >
          <br />
          <React.Fragment
            key="0"
          >
              
          </React.Fragment>
          <React.Fragment
            key="1"
          >
              
          </React.Fragment>
          quote
          : 
          Man, I'm pretty.
        </React.Fragment>
      </React.Fragment>
      <br />
    </React.Fragment>
  </p>
  <p>
    Status: 
    APPROVED
  </p>
  <button
    className="btn ui-action"
    onClick={[Function]}
    type="button"
  >
    Approve
  </button>
</React.Fragment>
`;

exports[`(Component) Rectification details should render when loading correctly 1`] = `undefined`;

exports[`(Component) Rectification details should render when not loading correctly 1`] = `
<React.Fragment>
  <p>
    <strong>
      Created at:
    </strong>
     
  </p>
  <p>
    <strong>
      Current data:
    </strong>
  </p>
  <p>
    <React.Fragment
      key="0"
    >
      firstName
      : 
      Johnny
      <br />
    </React.Fragment>
    <React.Fragment
      key="1"
    >
      email
      : 
      jb@coolcartoons.io
      <br />
    </React.Fragment>
    <React.Fragment
      key="2"
    >
      someKey
      : 
      Some data goes here
      <br />
    </React.Fragment>
  </p>
  <p>
    <strong>
      Updates:
    </strong>
  </p>
  <p>
    <React.Fragment
      key="0"
    >
      lastName
      : 
      Bravo
      <br />
    </React.Fragment>
  </p>
  <p>
    Status: 
    PENDING
  </p>
  <button
    className="btn ui-action"
    onClick={[Function]}
    type="button"
  >
    Approve
  </button>
</React.Fragment>
`;
